Detectors/MUON/MCH/Align/src/AlignmentSpec.cxx

```diff
@@ -0,0 +1,1096 @@
+    mAlign.SetAllowedVariation(2, 0.002);
+    mAlign.SetAllowedVariation(3, 2.0);
+
+    // Fix chambers
+    auto chambers = ic.options().get<string>("fix-chamber");
+    for (int i = 0; i < chambers.length(); ++i) {
+      if (chambers[i] == ',') {
+        continue;
+      }
+      int chamber = chambers[i] - '0';
+      LOG(info) << Form("%s%d", "Fixing chamber: ", chamber);
+      mAlign.FixChamber(chamber);
+    }
+
+    doMatched = ic.options().get<bool>("matched");
+    outFileName = ic.options().get<string>("output");
+    readFromRec = ic.options().get<bool>("use-record");
+
+    if (readFromRec) {
+      mAlign.SetReadOnly();
+      LOG(info) << "Reading records as input";
+    }
+    mAlign.init();
+
+    ic.services().get<CallbackService>().set<CallbackService::Id::Stop>([this]() {
+      LOG(info) << "Alignment duration = " << mElapsedTime.count() << " s";
+    });
+  }
+
+  //_________________________________________________________________________________________________
+  void finaliseCCDB(framework::ConcreteDataMatcher& matcher, void* obj)
+  {
+    /// finalize the track extrapolation setting
+    if (mCCDBRequest && base::GRPGeomHelper::instance().finaliseCCDB(matcher, obj)) {
+
+      if (matcher == framework::ConcreteDataMatcher("GLO", "GRPMAGFIELD", 0)) {
+        LOG(info) << "Loading magnetic field from CCDB";
+        auto grp = base::GRPGeomHelper::instance().getGRPMagField();
+        base::Propagator::initFieldFromGRP(grp);
+        TrackExtrap::setField();
+        mAlign.SetBFieldOn(TrackExtrap::isFieldON());
+        TrackExtrap::useExtrapV2();
+        trackFitter.initField(grp->getL3Current(), grp->getDipoleCurrent());
+      }
+
+      if (matcher == framework::ConcreteDataMatcher("GLO", "GEOMALIGN", 0)) {
+        LOG(info) << "Loading reference geometry from CCDB";
+        transformation = geo::transformationFromTGeoManager(*gGeoManager);
+        for (int i = 0; i < 156; i++) {
+          int iDEN = GetDetElemId(i);
+          transformRef[iDEN] = transformation(iDEN);
+        }
+      }
+    }
+  }
+
+  //_________________________________________________________________________________________________
+  void processWithMatching(vector<ROFRecord>& mchROFs, vector<TrackMCH>& mchTracks, vector<Cluster>& mchClusters,
+                           vector<dataformats::TrackMCHMID>& muonTracks)
+  {
+    // processing for each track
+    for (const auto& mchROF : mchROFs) {
+
+      for (int iMCHTrack = mchROF.getFirstIdx();
+           iMCHTrack <= mchROF.getLastIdx(); ++iMCHTrack) {
+        tracksAll += 1;
+        // MCH-MID matching
+        if (!FindMuon(iMCHTrack, muonTracks)) {
+          continue;
+        }
+        trackMCHMID += 1;
+
+        auto mchTrack = mchTracks.at(iMCHTrack);
+        int id_track = iMCHTrack;
+        int nb_clusters = mchTrack.getNClusters();
+
+        // Track selection, saving only tracks having exactly 10 clusters
+        if (nb_clusters <= 9) {
```

Comment:
What is the purpose of this selection? I does not select tracks with exactly 10 clusters as written in the comment. Because of the overlaps between DE, tracks can have up to 2 clusters per chamber. For the same reason, it does not select the number of fired chamber either, if this was the intent. Beside,  it will create holes in the acceptance (apart in the overlap regions) as soon as 1 DE is missing, which will likely impact the alignment of the DE in front of it.

Reply:
Yes you are right, there is no reason to select tracks having exactly 10 clusters; this comment was just simply outdated and I didn't pay attention to correct it. Our initial setup was to have tracks having at least 10 clusters, this is what meant to be this above selection.

Reply:
I am still a bit concerned about the risk of creating holes in the acceptance because of this cut but ok..